### PR TITLE
Registration for manual tournaments

### LINKED ISSUE #442: Screenshot collection
Acceptance criteria: collect attachments from DMs and forward to private channel

### PR DESCRIPTION
**Pull request by AlphaKretin**
## Description

Implements the process for a player to submit a deck and a tournament host to validate it. 

Includes the commands /deck and /open, as well as handling for modal and button interactions.

Closes #442
#443

## Checklist

- [x] I am following the [contributing guidelines](https://github.com/DawnbrandBots/emcee-tournament-bot/blob/master/.github/CONTRIBUTING.md).
- [ ] I have updated any relevant [user guides](https://github.com/DawnbrandBots/emcee-tournament-bot/blob/master/guides).
- [ ] I have updated any relevant [documentation](https://github.com/DawnbrandBots/emcee-tournament-bot/blob/master/docs).
- [ ] I have updated the [changelog](https://github.com/DawnbrandBots/emcee-tournament-bot/blob/master/changelog.md), or this change is not user-facing.

### CODE REVIEW

**Comment by AlphaKretin** [comment]
Some UX issues to iron out before we sign off on this, e.g. being able to more easily tell which decks have and haven't been looked at.

**Comment by AlphaKretin** [comment]
Realised a todo here which is actually respecting the capacity setting

**Comment by kevinlul** [comment]
https://github.com/DawnbrandBots/emcee-tournament-bot/blob/dce7329e4f7116a5bcbbbbd9ec12c0c6ea975cec/src/events/messageCreate.ts#L98

**Comment by AlphaKretin** [comment]
I've been testing the registration process and hit a roadblock. When clicking the button to register, Emcee tentatively registers the user in the database without a deck. However, in this process, we run into a NOT NULL constraint on `manual_participant.tournamentId`. The strange thing is, the participant is registered in the database and the ID is correctly taken from the associated tournament, but the code protests this error and crashes entirely. 

image

```
QueryFailedError: SqliteError: NOT NULL constraint failed: manual_participant.tournamentId
    at QueryFailedError.TypeORMError (C:\Users\lunal\Documents\Code\emcee-tournament-bot\node_modules\typeorm\error\src\error\TypeORMError.ts:7:9)
    at QueryFailedError (C:\Users\lunal\Documents\Code\emcee-tournament-bot\node_modules\typeorm\error\src\error\QueryFailedError.ts:9:9)
    at BetterSqlite3QueryRunner.<anonymous> (C:\Users\lunal\Documents\Code\emcee-tournament-bot\node_modules\typeorm\driver\src\driver\better-sqlite3\BetterSqlite3QueryRunner.ts:128:19)
    at step (C:\Users\lunal\Documents\Code\emcee-tournament-bot\node_modules\tslib\tslib.js:144:27)
    at Object.next (C:\Users\lunal\Documents\Code\emcee-tournament-bot\node_modules\tslib\tslib.js:125:57)
    at fulfilled (C:\Users\lunal\Documents\Code\emcee-tournament-bot\node_modules\tslib\tslib.js:115:62)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  query: 'UPDATE "manual_participant" SET "tournamentId" = ? WHERE ("tournamentId" = ? AND "discordId" = ?)',
  parameters: [ undefined, 1, '169299769695535105' ],
  driverError: SqliteError: NOT NULL constraint failed: manual_participant.tournamentId
      at BetterSqlite3QueryRunner.<anonymous> (C:\Users\lunal\Documents\Code\emcee-tournament-bot\node_modules\typeorm\driver\src\driver\better-sqlite3\BetterSqlite3QueryRunner.ts:109:38)
      at step (C:\Users\lunal\Documents\Code\emcee-tournament-bot\node_modules\tslib\tslib.js:144:27)
      at Object.next (C:\Users\lunal\Documents\Code\emcee-tournament-bot\node_modules\tslib\tslib.js:125:57)
      at fulfilled (C:\Users\lunal\Documents\Code\emcee-tournament-bot\node_modules\tslib\tslib.js:115:62)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
    code: 'SQLITE_CONSTRAINT_NOTNULL'
  },
  code: 'SQLITE_CONSTRAINT_NOTNULL'
}```

**Comment by kevinlul** [comment]
cannot reproduce

**Comment by kevinlul** [comment]
Found failure in duplicate  INSERT INTO "manual_deck_submission"

**Comment by kevinlul** [comment]
Can request a friend code via modal on button press before any part of the DM workflow even begins

**Comment by AlphaKretin** [comment]
Sidebar: At some point we'll have to figure out persistence through restarts and how to handle timeouts on the collectors (though they could probably go a lot longer than the placeholders filled in here)

File: src/slash/open.ts [hunk 1]
```diff
@@ -166,18 +182,22 @@ export class RegisterButtonHandler implements ButtonClickHandler {
...
-		const actionRow = new ActionRowBuilder<TextInputBuilder>().addComponents(deckLabelInput);
-		modal.addComponents(actionRow);
-		await interaction.showModal(modal);
+		if (tournament.requireFriendCode) {
+			const modal = new ModalBuilder().setCustomId("registerModal").setTitle(`Register for ${tournament.name}`);
+			const deckLabelInput = new TextInputBuilder()
+				.setCustomId("friendCode")
+				.setLabel("Master Duel Friend Code")
+				.setStyle(TextInputStyle.Short)
+				.setRequired(tournament.requireFriendCode)
```
Comment (AlphaKretin) [inline src/slash/open.ts:191]: If we're only asking for the friend code when required (which is a fair direction to take it), this should just be `true`

**Comment by kevinlul** [comment]
I would like to go with something like this so things stay focused and components are completed and polished piecemeal

**Pull request merged**